point_viewer_grpc/src/service.rs

```diff
@@ -245,19 +255,52 @@ impl proto_grpc::Octree for OctreeService {
     fn get_all_points(
         &self,
         ctx: RpcContext,
-        _req: proto::GetAllPointsRequest,
+        req: proto::GetAllPointsRequest,
         resp: ServerStreamingSink<proto::PointsReply>,
     ) {
-        stream_points_back_to_sink(OctreeQuery::FullPointcloud, self.octree.clone(), &ctx, resp)
+        stream_points_back_to_sink(
+            OctreeQuery::FullPointcloud,
+            Arc::clone(&self.get_service_meta(&req.octree_id)),
+            &ctx,
+            resp,
+        )
     }
 }
 
-pub fn start_grpc_server(octree_directory: &Path, host: &str, port: u16) -> Server {
+impl OctreeService {
+    pub fn get_service_meta(&self, octree_id: &str) -> Arc<OctreeServiceMeta> {
+        let octree_id = octree_id.to_string();
+        let mut cache = self.meta_cache.lock().unwrap();
+        match cache.get(&octree_id) {
+            None => {
+                let octree = self
+                    .factory
+                    .generate_octree(self.location.join(&octree_id).to_string_lossy())
+                    .unwrap();
+                let meta = octree.data_provider.meta_proto().unwrap();
+                let service_meta = Arc::new(OctreeServiceMeta { octree, meta });
+                cache.put(octree_id, Arc::clone(&service_meta));
+                service_meta
+            }
+            Some(service_meta) => Arc::clone(service_meta),
+        }
+    }
+}
+
+pub fn start_grpc_server(
+    host: &str,
+    port: u16,
+    location: impl Into<PathBuf>,
+    factory: OctreeFactory,
+) -> Server {
     let env = Arc::new(Environment::new(1));
-    let meta = read_meta_proto(octree_directory).unwrap();
-    let octree = Arc::from(octree_from_directory(octree_directory).unwrap());
+    let meta_cache = Arc::new(Mutex::new(LruCache::new(10)));
```

Comment:
could it be useful to set the cache size with an optional command line argument (taking a default parameter)? see for example sdl_viewer.

Reply:
I changed the lru cache to a hash map, since we don't have too many octrees at the moment, so setting the size currently is not relevant any more...